walkeeper/src/timeline.rs

```diff
@@ -410,8 +417,57 @@ impl Timeline {
         Ok(rmsg)
     }
 
-    pub fn get_info(&self) -> SafeKeeperState {
-        self.mutex.lock().unwrap().sk.s.clone()
+    pub fn get_state(&self) -> (SafekeeperMemState, SafeKeeperState) {
+        let shared_state = self.mutex.lock().unwrap();
+        (shared_state.sk.inmem.clone(), shared_state.sk.s.clone())
+    }
+
+    /// Prepare public safekeeper info for reporting.
+    pub fn get_public_info(&self) -> SafekeeperInfo {
+        let shared_state = self.mutex.lock().unwrap();
+        SafekeeperInfo {
+            last_log_term: Some(shared_state.sk.get_epoch()),
+            flush_lsn: Some(shared_state.sk.wal_store.flush_lsn()),
+            // note: this value is not flushed to control file yet and can be lost
+            commit_lsn: Some(shared_state.sk.inmem.commit_lsn),
+            s3_wal_lsn: Some(shared_state.sk.inmem.s3_wal_lsn),
+            // TODO: rework feedbacks to avoid max here
+            remote_consistent_lsn: Some(max(
+                shared_state.get_replicas_state().remote_consistent_lsn,
+                shared_state.sk.inmem.remote_consistent_lsn,
+            )),
+            peer_horizon_lsn: Some(shared_state.sk.inmem.peer_horizon_lsn),
+        }
+    }
+
+    /// Update timeline state with peer safekeeper data.
+    pub fn record_safekeeper_info(&self, sk_info: &SafekeeperInfo, _sk_id: ZNodeId) {
+        let mut shared_state = self.mutex.lock().unwrap();
+        // Note: the check is too restrictive, generally we can update local
+        // commit_lsn if our history matches (is part of) history of advanced
+        // commit_lsn provider.
+        if let (Some(commit_lsn), Some(last_log_term)) = (sk_info.commit_lsn, sk_info.last_log_term)
+        {
+            if last_log_term == shared_state.sk.get_epoch() {
+                let local_commit_lsn = min(commit_lsn, shared_state.sk.wal_store.flush_lsn());
+                shared_state.sk.inmem.commit_lsn =
```

Comment:
I'd update only `global_commit_lsn` here, and waited until it propagates to `commit_lsn` by `SafeKeeper::update_commit_lsn`. Also note that we are calling `notify_wal_senders(commit_lsn)` only inside `process_msg`.

I'm also not really sure if we need to update local commit_lsn from other peers, because it updates just fine when compute is alive. Is there anything we need to do with commit_lsn when compute is not alive?

Reply:
> I'd update only global_commit_lsn here,

Makes sense.

> I'm also not really sure if we need to update local commit_lsn from other peers

Yes, currently we don't have to. In perspective, it would be useful as sync-safekeepers replacement though.